runtime/src/main/java/org/corfudb/runtime/collections/ImmutableCorfuTable.java

```diff
@@ -0,0 +1,386 @@
+                .getOrElse(secondaryIndexesWrapper);
+
+        newSecondaryIndexesWrapper = newSecondaryIndexesWrapper.mapSecondaryIndexes(key, value);
+        return new ImmutableCorfuTable<>(mainMap.put(key, value), newSecondaryIndexesWrapper);
+    }
+
+    /**
+     * Delete a key-value pair, updating secondary indexes if applicable.
+     * @param key The key to remove.
+     * @return An ImmutableCorfuTable where the mapping given from the
+     * provided key is removed.
+     */
+    public ImmutableCorfuTable<K, V> remove(@Nonnull K key) {
+        SecondaryIndexesWrapper<K, V> newSecondaryIndexesWrapper = mainMap
+                .get(key)
+                .map(prev -> secondaryIndexesWrapper.unmapSecondaryIndexes(key, prev))
+                .getOrElse(secondaryIndexesWrapper);
+
+        return new ImmutableCorfuTable<>(mainMap.remove(key), newSecondaryIndexesWrapper);
+    }
+
+    /**
+     * Compute the size of this ImmutableCorfuTable.
+     * @return The number of key-value pairs in this ImmutableCorfuTable.
+     */
+    public int size() {
+        return mainMap.size();
+    }
+
+    /**
+     * Returns true if a mapping for the specified key exists.
+     * @param key The key whose presence is to be tested.
+     * @return True if and only if a mapping for the provided key exists.
+     */
+    public boolean containsKey(@Nonnull K key) {
+        return mainMap.containsKey(key);
+    }
+
+    /**
+     * Removes all of the key-value pair mappings.
+     * @return An ImmutableCorfuTable with all key-value mappings removed.
+     */
+    public ImmutableCorfuTable<K, V> clear() {
+        return new ImmutableCorfuTable<>(
+                HashMap.empty(),
+                secondaryIndexesWrapper.clear()
+        );
+    }
+
+    /**
+     * Returns a set of all keys present.
+     * @return A set containing all keys present in this ImmutableCorfuTable.
+     */
+    public java.util.Set<K> keySet() {
+        return mainMap.keySet().toJavaSet();
+    }
+
+    /**
+     * Returns the key-value mappings through the Java stream interface.
+     * @return A stream containing all key-value mappings in this ImmutableCorfuTable.
+     */
+    public Stream<java.util.Map.Entry<K, V>> entryStream() {
+        return StreamSupport.stream(spliterator(mainMap), true);
+    }
+
+    /**
+     * Get a mapping using the specified index function.
+     * @param indexName Name of the secondary index to query.
+     * @param indexKey The index key used to query the secondary index
+     * @param <I> The type of the index key.
+     * @return An Iterable of map entries satisfying this index query.
+     */
+    public <I> Iterable<java.util.Map.Entry<K, V>> getByIndex(@Nonnull final Index.Name indexName, I indexKey) {
+        return secondaryIndexesWrapper.contains(indexName.get(), indexKey)
+                .<Iterable<java.util.Map.Entry<K, V>>>map(IterableWrapper::new)
+                .getOrElse(Collections.emptySet());
+    }
+
```

Comment:
I know you haven't introduced this, but I think we should not allow for null-queries. 

Since multiple values can map to null, It is ambiguous and I don't think application code should use null for queries. Preventing null from the secondary index will also save. No null index = less memory usage.

Reply:
I want to maintain the same semantics for both CorfuTable implementations. I can address this in a subsequent PR.